dali/kernels/imgproc/convolution/convolution_gpu.h

```diff
@@ -0,0 +1,215 @@
+ * To the convolution we pass: alpha = `scale`, beta = 0 and D = C = `out`
+ * - we're intersted the multiplication only, the addition is "in-place" (to a zeroed matrix C)
+ * A and B correspond to `in` and `window` (order depends whether it's inner
+ * or outer convolution), effectively calculating D = alpha * A * B.
+ */
+template <typename Out, typename In, typename W, int ndim, int axis, bool has_channels = true>
+struct ConvolutionGpu {
+  KernelRequirements Setup(KernelContext& ctx, const TensorListShape<ndim>& in_shape,
+                           const TensorListShape<1>& window_size) {
+    KernelRequirements req;
+    ScratchpadEstimator se;
+    DALI_ENFORCE(
+        in_shape.size() == window_size.size(),
+        make_string("Provided input shape and window sizes do not mach in number of samples: ",
+                    in_shape.size(), " vs ", window_size.size(), "."));
+    int num_samples = in_shape.size();
+    for (int i = 0; i < num_samples; i++) {
+      int num_channels = has_channels ? in_shape[i][ndim - 1] : 1;
+      DALI_ENFORCE(window_size[i][0] % 2 == 1,
+                   make_string("Kernel window should have odd length, got: ", window_size,
+                               " for sample ", i, "."));
+
+      DALI_ENFORCE((window_size[i][0] / 2) * num_channels < kMaxRadiusSpan,
+                   make_string("Kernel window too big for sample ", i, "."));
+    }
+    se.add<W>(AllocType::Host, num_samples * kWindowCopyBufferSize);
+    se.add<W>(AllocType::GPU, num_samples * kWindowCopyBufferSize);
+    req.scratch_sizes = se.sizes;
+    req.output_shapes.push_back(in_shape);
+    return req;
+  }
+
+  void Run(KernelContext& ctx, const TensorListView<StorageGPU, Out, ndim> out,
+           const TensorListView<StorageGPU, const In, ndim>& in,
+           const TensorListView<StorageCPU, const W, 1>& window, W scale = 1) {
+    int num_samples = in.size();
+    auto* window_tmp_buffer_host =
+        ctx.scratchpad->Allocate<W>(AllocType::Host, num_samples * kWindowCopyBufferSize);
+    auto* window_tmp_buffer_gpu =
+        ctx.scratchpad->Allocate<W>(AllocType::GPU, num_samples * kWindowCopyBufferSize);
+
+    // Pad and align windows in tmp memory, transfer the aligned windows to GPU
+    FillAlignedWindow(window_tmp_buffer_host, window, in.shape);
+    cudaMemcpyAsync(window_tmp_buffer_gpu, window_tmp_buffer_host,
+                    sizeof(W) * num_samples * kWindowCopyBufferSize, cudaMemcpyHostToDevice,
+                    ctx.gpu.stream);
+
+    Arguments args;
+    if (kIsInnerConv) {
+      // Inner (innermost) - repack arguments
+      for (int i = 0; i < num_samples; i++) {
+        cutlass::Array<int, 2> size;
+        auto sample_shape = in.tensor_shape(i);
+        int num_channels = has_channels ? sample_shape[ndim - 1] : 1;
+        // height
+        size[0] = volume(sample_shape.begin(), sample_shape.begin() + axis);
+        // width
+        size[1] = sample_shape[ndim - has_channels - 1];
```

Comment:
It's not really horizontal. It looks like a matrix width from CUTLASS point of view, but it doesn't need to correspond to geometric width. I think it's possible to use it with 1D input. The outer volume should be 1, so in theory it should work...
@klecki Correct me if I'm wrong.

Reply:
Yes, the matrix is Height x Width, and for the case here it's the innermost dimension of the input, for 1D, the height is just 1. Actually, this probably could be the `axis` here, it may be a bit clearer that way.